File: Irrgarten_ruby/Game.rb
#encoding:utf-8
module Irrgarten
    # The Game class represents a game of Irrgarten.
    class Game
        @@MAX_ROUNDS = 10

        # Initializes a new instance of the Game class.
        # @param nplayers [Integer] The number of players in the game.
        def initialize(nplayers)
            @players = Array.new(nplayers)
            nplayers.times { |i|
                @players[i] = Player.new(i, Dice.random_intelligence, Dice.random_strength)
            }
            @current_player_index = Dice.who_starts(nplayers)
            @current_player = @players[@current_player_index]
            @monsters = Array.new(0)
            @log = ""
            configure_labyrinth
        end

        # Checks if the game has finished.
        # @return [Boolean] True if the game has finished, false otherwise.
        def finished
            @labyrinth.have_a_winner
        end

        # Moves to the next step in the game.
        # @param preferred_direction [Symbol] The preferred direction to move.
        # @return [Boolean] True if the game has finished, false otherwise.
        def next_step(preferred_direction)
            @log=""
            dead=@current_player.dead
            if !dead
                direction = actual_direction(preferred_direction)
                if direction != preferred_direction
                    log_player_no_orders
                end

                monster = @labyrinth.put_player(direction, @current_player)

                if monster.nil?
                    log_no_monster
                else
                    winner=combat(monster)
                    manage_reward(winner)
                end
            else
                manage_resurrection
            end

            end_game=finished
            if !end_game
                next_player
            end

            end_game
        end

        # Gets the current state of the game.
        # @return [GameState] The current state of the game.
        def game_state
            labyrinth_s = @labyrinth.to_s
            players_s = ""
            @players.each do |player|
                players_s += player.to_s + "\n"
            end
            monsters_s = ""
            @monsters.each do |monster|
                monsters_s += monster.to_s + "\n"
            end
            current_p = @current_player_index
            winner = self.finished
            state = GameState.new(labyrinth_s, players_s, monsters_s, current_p, winner, @log)
            state
        end

        private

        # Configures the labyrinth for the game.
        def configure_labyrinth
            n_rows = 5
            n_cols = 5
            exit_row = 1
            exit_col = 1
            @labyrinth = Labyrinth.new(n_rows, n_cols, exit_row, exit_col)

            @labyrinth.add_block(Orientation::HORIZONTAL, 0, 0, 2)
            @labyrinth.add_block(Orientation::VERTICAL, 0, 1, 2)

            goblin = Monster.new("Goblin", Dice.random_intelligence, Dice.random_strength)
            @labyrinth.add_monster(2, 2, goblin)
            @monsters << goblin

            piglin = Monster.new("Piglin", Dice.random_intelligence, Dice.random_strength)
            @labyrinth.add_monster(4, 4, piglin)
            @monsters << piglin

            zombie = Monster.new("Zombie", Dice.random_intelligence, Dice.random_strength)
            @labyrinth.add_monster(4, 0, zombie)
            @monsters << zombie

            skeleton = Monster.new("Skeleton", Dice.random_intelligence, Dice.random_strength)
            @labyrinth.add_monster(0, 4, skeleton)
            @monsters << skeleton

            @labyrinth.spread_players(@players)
        end

        # Moves to the next player in the game.
        def next_player
            if (@current_player_index == @players.length - 1)
                @current_player_index = 0
            else
                @current_player_index = @current_player_index + 1
            end

            @current_player = @players[@current_player_index]
        end

        # Gets the actual direction to move.
        # @param preferred_direction [Symbol] The preferred direction to move.
        # @return [Symbol] The actual direction to move.
        def actual_direction(preferred_direction)
            current_row=@current_player.row
            current_col=@current_player.col
            valid_moves = @labyrinth.valid_moves(current_row, current_col)
            @current_player.move(preferred_direction, valid_moves)
        end

        # Manages the combat between the player and a monster.
        # @param monster [Monster] The monster to fight.
        # @return [Boolean] True if the player has won, false if the monster has won.
        # @note The combat is resolved in turns, with the player attacking first.
        def combat(monster)
            rounds = 0
            winner = GameCharacter::PLAYER
            player_attack= @current_player.attack
            lose = monster.defend(player_attack)
            while !lose && rounds < @@MAX_ROUNDS
                winner = GameCharacter::MONSTER
                rounds+=1
                monster_attack = monster.attack
                lose = @current_player.defend(monster_attack)
                if !lose
                    player_attack = @current_player.attack
                    winner = GameCharacter::PLAYER
                    lose = monster.defend(player_attack)
                end

            end
            log_rounds(rounds, @@MAX_ROUNDS)
            winner
        end

        # Manages the reward of the winner of a combat.
        # @param winner [Symbol] The winner of the combat.
        # @note The winner of the combat receives a reward.
        # @note The reward is a random weapon or shield.
        def manage_reward(winner)
            if winner==GameCharacter::PLAYER
                @current_player.receive_reward
                log_player_won
            else
                log_monster_won
            end
        end

        # Manages the resurrection of the current player.
        # @note The player has a chance to resurrect.
        def manage_resurrection
            resurrect = Dice.resurrect_player
            if resurrect
                @current_player.resurrect
                log_resurrected
            else
                log_player_skip_turn
            end
        end

        # Logs that the current player has won the game.
        def log_player_won
            @log += "Player number " + @current_player.number.to_s + " has won. \n"
        end

        # Logs that the monster has won the game.
        def log_monster_won
            @log += "The monster has won \n"
        end

        # Logs that the current player has been resurrected.
        def log_resurrected
            @log += "The player " + @current_player.number.to_s + " has resurrected.\n"
        end

        # Logs that the current player has skipped a turn because they were dead.
        def log_player_skip_turn
            @log += "Player number " + @current_player.number.to_s + " has lost the turn " +
                "because they were dead. \n"
        end

        # Logs that the current player couldn't follow the instructions from the human player.
        def log_player_no_orders
            @log += "Player number " + @current_player.number.to_s + " couldn't follow " +
                "the instructions from the human player. \n"
        end

        # Logs that the current player couldn't move or moved to an empty box.
        def log_no_monster
            @log += "Player number " + @current_player.number.to_s + " couldn't move or " +
                "moved to an empty box. \n"
        end

        # Logs the number of rounds passed in the game.
        # @param rounds [Integer] The number of rounds passed.
        # @param max [Integer] The maximum number of rounds.
        def log_rounds(rounds, max)
            @log += "Rounds passed " + rounds.to_s + " / " + max.to_s + "\n"
        end
    end
end

File: Irrgarten/src/irrgarten/Game.java
package irrgarten;
import java.util.ArrayList;

/**
 * The Game class represents a game of Irrgarten.
 * It manages the players, monsters, labyrinth, and game state.
 */
public class Game {
    /**
     * Max rounds of the game
     */
    public static final int MAX_ROUNDS = 10;
    /**
     * The index of the current player
     */
    private int currentPlayerIndex;
    /**
     * Log of the game
     */
    private String log;

    /**
     * The player who is playing at the moment
     */
    private Player currentPlayer;
    
    /**
     * Labyrinth of the game
     */
    private Labyrinth labyrinth;
    /**
     * Monsters in the game at the moment
     */
    private ArrayList<Monster> monsters;
    
    /**
     * Players in the game at the moment
     */
    private ArrayList<Player> players;

    /**
     * Constructs a new Game object with the specified number of players.
     * @param nplayers The number of players in the game.
     */
    public Game(int nplayers){
        this.log = "";
        this.players = new ArrayList<Player>(nplayers);
        for (int i = 0; i < nplayers; i++){
            Player newPlayer = new Player((char)(i+'0'), 
                            Dice.randomIntelligence(), Dice.randomStrength());
            this.players.add(newPlayer);
        }
        this.currentPlayerIndex = Dice.whoStarts(nplayers);
        this.currentPlayer = this.players.get(currentPlayerIndex);
        this.monsters = new ArrayList<Monster> (0);
        this.configureLabyrinth();
    }

    /**
     * Checks if the game has finished.
     *
     * @return true if the game has finished, false otherwise.
     */
    public boolean finished(){
        return this.labyrinth.haveAWinner();
    }

    /**
     * Advances the game one step.
     * The player moves in the preferred direction and fights the monster in the new position.
     * If the player dies, it may resurrect.
     * If the game has finished, it advances to the next player.
     * @param preferredDirection The preferred direction of the player.
     * @return true if the game has finished, false otherwise.
     */
    public boolean nextStep(Directions preferredDirection) {
        this.log = ""; // Clear the log of the game because of a new step
        boolean dead = this.currentPlayer.dead();

        if(!dead) {
            Directions direction = this.actualDirection(preferredDirection);
            if(direction!=preferredDirection)
                this.logPlayerNoOrders();
            Monster monster = this.labyrinth.putPlayer(direction, this.currentPlayer);
            
            if(monster==null) // There is no monster
                this.logNoMonster();
            else{
                GameCharacter winner = this.combat(monster);
                this.manageReward(winner);
            }
        }
        else{
            this.manageResurrection();
        }
        
        boolean endGame = this.finished();
        if(!endGame){
            this.nextPlayer();
        }
        
        return endGame;
    }

    /**
     * Gets the current game state.
     *
     * @return The current game state.
     */
    public GameState getGameState() {
        String labyrinth_s = this.labyrinth.toString();
        String players_s = this.players.toString();
        String monsters_s = this.monsters.toString();
        int currentP = this.currentPlayerIndex;
        boolean winner = this.finished();

        GameState state = new GameState(labyrinth_s, players_s, monsters_s, currentP, winner, this.log);
        return state;
    }

    /**
     * Configures the labyrinth with the specified parameters.
     */
    private void configureLabyrinth(){
        int nRows = 5, nCols = 5;
        int exitRow = 1, exitCol = 1;
        
        this.labyrinth = new Labyrinth(nRows,nCols,exitRow, exitCol);
        this.labyrinth.addBlock(Orientation.HORIZONTAL, 0, 0, 2);

        Monster goblin = new Monster ("Goblin", Dice.randomIntelligence(), Dice.randomStrength());
        this.labyrinth.addMonster(2,2,goblin);
        this.monsters.add(goblin);
        
        Monster piglin = new Monster ("Piglin", Dice.randomIntelligence(), Dice.randomStrength());
        this.labyrinth.addMonster(4, 4, piglin);
        this.monsters.add(piglin);
        
        Monster zombie = new Monster ("Zombie", Dice.randomIntelligence(), Dice.randomStrength());
        this.labyrinth.addMonster(4, 0, zombie);
        this.monsters.add(zombie);
        
        Monster skeleton = new Monster ("Skeleton", Dice.randomIntelligence(), Dice.randomStrength());
        this.labyrinth.addMonster(0, 4, skeleton);
        this.monsters.add(skeleton);
        
        this.labyrinth.spreadPlayers(this.players);
    }

    /**
     * Advances the game to the next player.
     */
    private void nextPlayer(){
        int nPlayers = this.players.size();
        if (this.currentPlayerIndex == nPlayers-1){
            this.currentPlayerIndex = 0;
        } else {
            this.currentPlayerIndex++;
        }

        this.currentPlayer = this.players.get(this.currentPlayerIndex);
    }

    /**
     * Returns the actual direction of the player.
     * If the preferred direction is not valid, it returns the first valid direction.
     * @param preferredDirection The preferred direction of the player.
     * @return The actual direction of the player.
     */
    private Directions actualDirection(Directions preferredDirection) {
        int currentRow = this.currentPlayer.getRow();
        int currentCol = this.currentPlayer.getCol();
        ArrayList<Directions> validMoves = 
                            this.labyrinth.validMoves(currentRow, currentCol);
        Directions output = 
                        this.currentPlayer.move(preferredDirection, validMoves);
        return output;
    }

    /**
     * Simulates a combat between the player and the monster.
     * The combat ends when one of the characters dies or the maximum number of rounds is reached.
     * @param monster The monster to fight.
     * @return The winner of the combat.
     */
    private GameCharacter combat(Monster monster){
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        float playerAttack = this.currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        float monsterAttack;
        
        while(!lose && (rounds < Game.MAX_ROUNDS)){
            winner = GameCharacter.MONSTER;
            rounds++;
            monsterAttack = monster.attack();
            lose = this.currentPlayer.defend(monsterAttack);
            if(!lose){
                playerAttack = this.currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack); 
            }
        }
        
        this.logRounds(rounds, Game.MAX_ROUNDS);
        return winner;
    }

    /**
     * Manages the reward of the combat.
     * If the player wins, it receives a reward.
     * @param winner The winner of the combat.
     * @return The winner of the combat.
     */
    private void manageReward(GameCharacter winner) {
        if (winner == GameCharacter.PLAYER){
            this.currentPlayer.receiveReward();
            this.logPlayerWon();
        }
        else
            this.logMonsterWon();
    }

    /**
     * Manages the resurrection of the player.
     * If the player doesn't resurrect, it loses the turn.
     * @return The winner of the combat.
     */
    private void manageResurrection() {
        boolean resurrect = Dice.resurrectPlayer();
        if (resurrect){
            this.currentPlayer.resurrect();
            this.currentPlayer = new FuzzyPlayer(this.currentPlayer);
            this.players.set(this.currentPlayerIndex, this.currentPlayer);
            this.labyrinth.updatePlayer(currentPlayer);
            this.logResurrected();
        }
        else
            this.logPlayerSkipTurn();
    }

    /**
     * Logs the player's victory.
     */
    private void logPlayerWon(){
        this.log += "Player number " + this.currentPlayer.getNumber() + 
                    " has won the combat. \n";
    }

    /**
     * Logs the monster's victory.
     */
    private void logMonsterWon(){
        this.log += "The monster has won the combat\n";
    }
    
    /**
     * Logs the resurrection of the player.
     */
    private void logResurrected(){
        this.log += "The player" + this.currentPlayer.getNumber() + 
                    " has resurrected.\n";
    }
    
    /**
     * Logs the player's turn skip.
     */
    private void logPlayerSkipTurn(){
        this.log += "Player number " + this.currentPlayer.getNumber() + 
                    " has lost the turn because he was dead. \n";
    }
    
    /**
     * Logs the player's turn skip.
     */
    private void logPlayerNoOrders(){
        this.log += "Player number " + this.currentPlayer.getNumber() + 
                    " moved somewhere else.\n"; 
    }
    
    /**
     * Logs the player's turn skip.
     */
    private void logNoMonster(){
        this.log += "Player number " + this.currentPlayer.getNumber() + 
                    " couldn't move or moved to an empty box.\n";
    }
    
    /**
     * Logs the number of rounds passed.
     */
    private void logRounds(int rounds, int max){
        this.log += "Rounds passed: " + rounds + " / " + max + "\n"; 
    }
}

